Meadowlark-js/backends/meadowlark-mongodb-backend/src/repository/Upsert.ts

```diff
@@ -3,64 +3,82 @@
 // The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 // See the LICENSE and NOTICES files in the project root for more information.
 
-import { Collection, ClientSession, MongoClient, WithId, FindOptions } from 'mongodb';
-import { UpsertResult, UpsertRequest, documentIdForSuperclassInfo, BlockingDocument } from '@edfi/meadowlark-core';
+/* eslint-disable no-underscore-dangle */
+
+import { Collection, ClientSession, MongoClient, WithId } from 'mongodb';
+import {
+  UpsertResult,
+  UpsertRequest,
+  documentIdForSuperclassInfo,
+  BlockingDocument,
+  DocumentUuid,
+} from '@edfi/meadowlark-core';
 import { Logger } from '@edfi/meadowlark-utilities';
 import { MeadowlarkDocument, meadowlarkDocumentFrom } from '../model/MeadowlarkDocument';
-import { getDocumentCollection, writeLockReferencedDocuments, asUpsert } from './Db';
+import { writeLockReferencedDocuments, asUpsert, limitFive, getDocumentCollection, onlyReturnDocumentUuid } from './Db';
 import { onlyDocumentsReferencing, validateReferences } from './ReferenceValidation';
 
-// MongoDB FindOption to return at most 5 documents
-const limitFive = (session: ClientSession): FindOptions => ({ limit: 5, session });
-
 const moduleName: string = 'mongodb.repository.Upsert';
+
 export async function upsertDocumentTransaction(
-  { resourceInfo, documentInfo, documentUuidInserted, meadowlarkId, edfiDoc, validate, traceId, security }: UpsertRequest,
-  client: MongoClient,
+  {
+    resourceInfo,
+    documentInfo,
+    documentUuidForInsert,
+    meadowlarkId,
+    edfiDoc,
+    validateDocumentReferencesExist,
+    traceId,
+    security,
+  }: UpsertRequest,
+  mongoCollection: Collection<MeadowlarkDocument>,
   session: ClientSession,
+  documentFromUpdate?: MeadowlarkDocument,
 ): Promise<UpsertResult> {
-  const mongoCollection: Collection<MeadowlarkDocument> = getDocumentCollection(client);
-  let upsertResult: UpsertResult = { response: 'UNKNOWN_FAILURE' };
-  let documentUuid = documentUuidInserted;
+  // Check whether this document exists in the db
+  const existingDocument: WithId<MeadowlarkDocument> | null = await mongoCollection.findOne(
+    { _id: meadowlarkId },
+    onlyReturnDocumentUuid(session),
+  );
+
+  // the documentUuid of the existing document if this is an update, or null if this is an insert
+  const existingDocumentUuid: DocumentUuid | null = existingDocument?.documentUuid ?? null;
```

Comment:
This is the other documentUuid at play, the (optional) one where this is an update condition.

Reply:
Note there is no reason to make this db call if this call is coming from updateByDocumentId. I ran out of time, and a minor optimization since it's probably not the common case.